language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,436 +110,178 @@ impl AbstractState {
-                                .union(&nonce_set)
-                                .cloned()
-                                .collect::<BTreeSet<_>>();
-                            if !borrowed_nonce_set_update.is_empty() {
-                                new_index_to_nonce_set
-                                    .insert(idx.clone(), borrowed_nonce_set_update);
-                            }
-                        } else {
-                            new_index_to_nonce_set.insert(idx.clone(), borrowed_nonce_set.clone());
-                        }
-                    }
-                    if !new_index_to_nonce_set.is_empty() {
-                        new_borrows.insert(
-                            *src_nonce,
-                            BorrowInfo::FieldsBorrowedBy(new_index_to_nonce_set),
-                        );
-                    }
-                }
-            }
-        }
+    /// returns all borrows of a global
+    pub fn global_borrows(&self, idx: StructDefinitionIndex) -> BTreeSet<Nonce> {
+        self.borrow_graph
+            .consistent_borrows(self.frame_root, LabelElem::Global(idx))
+    }
 
-        self.locals = new_locals;
-        self.globals = new_globals;
-        self.borrows = new_borrows;
-        self.partition.remove_nonce(nonce);
+    /// returns all borrows of a nonce that might be consistent with a field
+    pub fn field_borrows(&self, nonce: Nonce, idx: FieldDefinitionIndex) -> BTreeSet<Nonce> {
+        self.borrow_graph
+            .consistent_borrows(nonce, LabelElem::Field(idx))
     }
 
-    /// checks if there are any pending borrows on value
-    pub fn is_full(&self, value: &AbstractValue) -> bool {
-        match value {
-            AbstractValue::Reference(nonce) => !self.borrows.contains_key(&nonce),
-            AbstractValue::Value(_, nonce_set) => nonce_set.is_empty(),
-        }
+    /// checks if a nonce is borrowed
+    pub fn is_nonce_borrowed(&self, nonce: Nonce) -> bool {
+        !self.borrow_graph.all_borrows(nonce).is_empty()
     }
 
-    /// returns the set of nonces borrowing from nonce that might alias some idx-extension of nonce
-    pub fn borrowed_nonces_for_field(
-        &self,
-        idx: FieldDefinitionIndex,
-        nonce: Nonce,
-    ) -> BTreeSet<Nonce> {
-        if self.borrows.contains_key(&nonce) {
-            match &self.borrows[&nonce] {
-                BorrowInfo::BorrowedBy(borrowed_by) => borrowed_by.clone(),
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    if fields_borrowed_by.contains_key(&idx) {
-                        fields_borrowed_by[&idx].clone()
-                    } else {
-                        BTreeSet::new()
-                    }
-                }
-            }
-        } else {
-            BTreeSet::new()
-        }
+    /// checks if a nonce is borrowed via an edge with the empty label
+    pub fn is_nonce_nil_borrowed(&self, nonce: Nonce) -> bool {
```

Comment:
"nil" is probably not such a great term to use here. I also find myself wondering if there can be more than one edge and if so, whether the other edges may have non empty labels. Also, is an empty label actually a label, or should one speak of an edge being "unlabelled"?

Reply:
A label is a sequence of names.  nil refers to the empty sequence.  All edges must be labeled by some sequence, empty or nonempty.

Do you have a suggestion to replace nil?

Reply:
I'm still not clear what the circumstances are that lead to an edge having an empty label. Perhaps a name will come to you if you explain that.

Reply:
The label on an edge represents a sequence of field lookups.  So, if there is an edge labeled p from n to n', it is saying that the p-extension of n is equal (if the edge is strong) or a prefix (if the edge is weak) of n'.  If n' is a copy of n, then the edge is strong and p is empty.  If n' is an unknown extension of n (say n was passed into a procedure and n' was returned), the the edge is weak and p is empty.

How about using empty instead of nil?

Reply:
I have removed this function.  It is not needed after my latest refactoring.